src/priv_appendable_data.rs

```diff
@@ -0,0 +1,418 @@
+use xor_name::XorName;
+use data::DataIdentifier;
+use error::RoutingError;
+use pub_appendable_data::{SERIALISED_APPENDED_DATA_SIZE, AppendedData, Filter};
+
+/// Maximum allowed size for a private appendable data to grow to
+pub const MAX_PRIV_APPENDABLE_DATA_SIZE_IN_BYTES: usize = 102400;
+
+/// Private appendable data.
+///
+/// These types may be stored unsigned with previous and current owner keys
+/// set to the same keys. Updates require a signature to validate.
+///
+/// Data can be appended by any key that is not excluded by the filter.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, RustcDecodable, RustcEncodable)]
+pub struct PrivAppendableData {
+    name: XorName,
+    version: u64,
+    current_owner_keys: Vec<PublicKey>,
+    previous_owner_keys: Vec<PublicKey>,
+    filter: Filter,
+    deleted_data: BTreeSet<AppendedData>,
+    previous_owner_signatures: Vec<Signature>, // All the above fields
+    data: BTreeSet<AppendedData>, // Unsigned
+}
+
+impl PrivAppendableData {
+    /// Creates a new `PubAppendableData` signed with `signing_key`.
+    pub fn new(name: XorName,
+               version: u64,
+               current_owner_keys: Vec<PublicKey>,
+               previous_owner_keys: Vec<PublicKey>,
+               filter: Filter,
+               signing_key: Option<&SecretKey>)
+               -> Result<PrivAppendableData, RoutingError> {
+
+        let mut priv_appendable_data = PrivAppendableData {
+            name: name,
+            version: version,
+            current_owner_keys: current_owner_keys,
+            previous_owner_keys: previous_owner_keys,
+            filter: filter,
+            deleted_data: BTreeSet::new(),
+            previous_owner_signatures: vec![],
+            data: BTreeSet::new(),
+        };
+
+        if let Some(key) = signing_key {
+            let _ = try!(priv_appendable_data.add_signature(key));
+        }
+        Ok(priv_appendable_data)
+    }
+
+    /// Deletes the given data item and returns `true` if it was there.
+    pub fn delete(&mut self, appended_data: AppendedData) -> bool {
+        let removed = self.data.remove(&appended_data);
+        if removed {
+            let _ = self.deleted_data.insert(appended_data);
+        }
+        removed
+    }
+
+    /// Inserts the given data item, or returns `false` if it cannot be added because it has
+    /// recently been deleted.
+    pub fn append(&mut self, appended_data: AppendedData) -> bool {
+        if self.deleted_data.contains(&appended_data) {
+            return false;
+        }
+        let signed_data = match serialise(&(appended_data.pointer(), appended_data.sign_key())) {
+            Ok(result) => result,
+            Err(_) => return false,
+        };
+        if sign::verify_detached(appended_data.signature(),
+                                 &signed_data,
+                                 appended_data.sign_key()) {
+            // PrivAppendableData shall only allows working with white list
```

Comment:
I think that should be checked against `AppendWrapper::Priv::sign_key` - the signing key of the private appended data item will be encrypted.
And I don't see why we shouldn't check that key against a black list instead of a white list.